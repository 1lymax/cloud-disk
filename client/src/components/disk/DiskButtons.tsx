import {Button} from "@mui/material";
import React, {FC, useState} from 'react';
import SyncIcon from '@mui/icons-material/Sync';
import CreateNewFolderIcon from "@mui/icons-material/CreateNewFolder";
import KeyboardBackspaceIcon from "@mui/icons-material/KeyboardBackspace";

import ModalCreateDir from "../UI/ModalCreateDir";
import LoadingButton from "@mui/lab/LoadingButton";
import {popDirStack} from "../../store/reducers/FileSlice";
import {useAppDispatch, useAppSelector} from "../../hooks/hooks";


interface DiskButtonsProps {
    refetch: () => void;
    isLoading: boolean;
}


const DiskButtons: FC <DiskButtonsProps> = ({refetch, isLoading}) => {
    const dispatch = useAppDispatch()
    const [modalOpen, setModalOpen] = useState<boolean>(false);
    const currentDir = useAppSelector(state => state.fileState.currentDir)


    const handleBackButton = () => {
        dispatch(popDirStack())
    };

    return (
        <>
            {currentDir &&
				<Button size="large"
						startIcon={<KeyboardBackspaceIcon/>}
						variant="text"
						sx={{textTransform: "none"}}
						onClick={handleBackButton}
				>
					Back
				</Button>
            }
            <Button size="large"
                    startIcon={<CreateNewFolderIcon/>}
                    variant="text"
                    sx={{textTransform: "none"}}
                    onClick={() => setModalOpen(!modalOpen)}
            >
                Create Dir
            </Button>
            <LoadingButton loading={isLoading} size="large"
                    startIcon={<SyncIcon/>}
                    variant="text"
                    sx={{textTransform: "none"}}
                    onClick={() => refetch()}
            >
                Refresh
            </LoadingButton>
            <ModalCreateDir currentDir={currentDir} open={modalOpen} setOpen={setModalOpen}/>

        </>
    );
};

export default DiskButtons;