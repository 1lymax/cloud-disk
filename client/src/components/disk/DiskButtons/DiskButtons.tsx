import {Button} from "@mui/material";
import React, {FC, useState} from 'react';
import SyncIcon from '@mui/icons-material/Sync';
import CreateNewFolderIcon from "@mui/icons-material/CreateNewFolder";
import KeyboardBackspaceIcon from "@mui/icons-material/KeyboardBackspace";

import ModalCreateDir from "../../UI/ModalCreateDir";
import LoadingButton from "@mui/lab/LoadingButton";
import {popDirStack} from "../../../store/reducers/FileSlice";
import {useAppDispatch, useAppSelector} from "../../../hooks/hooks";


interface DiskButtonsProps {
    refetch: () => void;
    isLoading: boolean;
}


const DiskButtons: FC<DiskButtonsProps> = ({refetch, isLoading}) => {
    const dispatch = useAppDispatch()
    const [modalOpen, setModalOpen] = useState<boolean>(false);
    const currentDir = useAppSelector(state => state.fileState.currentDir)


    const handleBackButton = () => {
        dispatch(popDirStack())
    };

    return (
        <div data-testid='diskButtons'>
            {currentDir &&
				<Button size="large"
						data-testid='backDir-button'
						startIcon={<KeyboardBackspaceIcon/>}
						variant="text"
						sx={{textTransform: "none"}}
						onClick={handleBackButton}
				>
					Back
				</Button>
            }
            <Button size="large"
                    data-testid='createDir-button'
                    startIcon={<CreateNewFolderIcon/>}
                    variant="text"
                    sx={{textTransform: "none"}}
                    onClick={() => setModalOpen(!modalOpen)}
            >
                Create Dir
            </Button>
            <LoadingButton loading={isLoading} size="large"
                           startIcon={<SyncIcon/>}
                           data-testid='refresh-button'
                           variant="text"
                           sx={{textTransform: "none"}}
                           onClick={() => refetch()}
            >
                Refresh
            </LoadingButton>
            <ModalCreateDir currentDir={currentDir} open={modalOpen} setOpen={setModalOpen}/>

        </div>
    );
};

export default DiskButtons;