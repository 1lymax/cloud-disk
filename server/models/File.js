const {Schema, model} = require('mongoose')


const File = new Schema({
	name: {type: String, required: true},
	type: {type: String, required: true},
	accessLink: {type: String},
	size: {type: Number, default: 0},
	path: {type: String, default: ''},
	user: {type: Schema.Types.ObjectId, ref: 'User'},
	parent: {type: Schema.Types.ObjectId, ref: 'File'},
	childs: [{type: Schema.Types.ObjectId, ref: 'File'}]
})

module.exports = model('File', File)