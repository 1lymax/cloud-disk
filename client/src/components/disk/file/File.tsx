import {useSnackbar} from "notistack";
import ShareIcon from '@mui/icons-material/Share';
import LoadingButton from "@mui/lab/LoadingButton";
import React, {FC, useEffect, useState} from 'react';
import FolderOpenIcon from '@mui/icons-material/FolderOpen';
import DeleteForeverIcon from '@mui/icons-material/DeleteForever';
import CloudDownloadIcon from '@mui/icons-material/CloudDownload';
import {IconButton, Tooltip, Typography} from "@mui/material";
import TextSnippetOutlinedIcon from '@mui/icons-material/TextSnippetOutlined';

import {IFile} from "../../../models/IFile";
import sizeFormat from "../../../utils/sizeFormat";
import {fileAPI} from "../../../services/FileService";
import {fileDownload} from "../../../actions/fileDownload";
import {getErrorMessage} from "../../../utils/getErrorMessage";
import {useAppDispatch, useAppSelector} from "../../../hooks/hooks";
import {popDirStack, pushDirStack, setCurrentDir} from "../../../store/reducers/FileSlice";

import classes from './File.module.scss'


interface FileProps {
    file: IFile;
    refetch: () => void;
}

const File: FC<FileProps> = ({file, refetch}) => {
    const dispatch = useAppDispatch()
    const {enqueueSnackbar} = useSnackbar()
    const fileView = useAppSelector(state => state.fileState.fileView)
    const currentDir = useAppSelector(state => state.fileState.currentDir)
    const [actionButtonsVisible, setActionButtonsVisible] = useState(false)
    const [deleteFile, {
        isLoading: deleteLoading,
        error: deleteError,
        isSuccess: deleteSuccess
    }] = fileAPI.useDeleteFileMutation()

    const downloadHandler = async (e: React.MouseEvent<HTMLButtonElement>) => {
        e.stopPropagation()
        fileDownload(file)
    }

    const handleClick = () => {
        if (file.type === 'dir') {
            if (currentDir === file._id) {
                dispatch(popDirStack());
            } else {
                dispatch(setCurrentDir(file._id ? file._id : ''));
                dispatch(pushDirStack(file._id));
            }
        }

    };

    const deleteHandler = () => {
        deleteFile(file._id)
    };

    useEffect(() => {
        if (deleteError) enqueueSnackbar(getErrorMessage(deleteError), {variant: "error"});
    }, [deleteError]);

    useEffect(() => {
        if (deleteSuccess) {
            enqueueSnackbar('File was deleted', {variant: "success"});
            refetch()
        }
    }, [deleteSuccess]);


    return (
        <div onClick={() => handleClick()}
             className={fileView === 'list' ? classes.list : classes.plate}
             onMouseEnter={() => setActionButtonsVisible(true)}
             onMouseLeave={() => setActionButtonsVisible(false)}
        >
            <div className={fileView === 'list' ? classes.list__icon : classes.plate__icon}>
                {file.type === "dir"
                    ? currentDir !== file._id &&
					<FolderOpenIcon
						className={fileView === 'list' ? classes.list__icon__folder : classes.plate__icon__folder}
						color="primary"/>
                    :
                    <TextSnippetOutlinedIcon
                        className={fileView === 'list' ? classes.list__icon__file : classes.plate__icon__file}
                    />
                }
            </div>
            <div className={fileView === 'list' ? classes.list__name : classes.plate__name}
                style={file.name.length > 23 ? {justifyContent: 'left'} : {}}
            >
                <Typography variant="subtitle1">
                    <Tooltip title={file.name.length > 23 ? file.name : ''}>
                        <Typography variant="subtitle2">
                            {file.name}
                        </Typography>
                    </Tooltip>
                </Typography>
            </div>
            <div className={fileView === 'list' ? classes.list__actions : classes.plate__actions}
                 style={!actionButtonsVisible ? {display: 'none'} : {}}>
                {file.type !== "dir" &&
					<IconButton color='secondary' onClick={(e) => downloadHandler(e)}>
						<CloudDownloadIcon/>
					</IconButton>
                }
                <LoadingButton loading={deleteLoading} variant='text' color='secondary' onClick={deleteHandler}>
                    <DeleteForeverIcon/>
                </LoadingButton>
                <IconButton color='secondary'>
                    <ShareIcon/>
                </IconButton>
            </div>
            <div className={fileView === 'list' ? classes.list__date : classes.plate__date}>
                <Tooltip title={`${file.date.slice(4, 16)} ${file.date.slice(16, 24)}`}>
                    <Typography variant="subtitle2">
                        {file.date.slice(4, 15)}
                    </Typography>
                </Tooltip>

            </div>
            <div className={fileView === 'list' ? classes.list__size : classes.plate__size}>
                <Typography variant="subtitle2">
                    {sizeFormat(file.size as number)}
                </Typography>
            </div>
        </div>
    );
};

export default File;