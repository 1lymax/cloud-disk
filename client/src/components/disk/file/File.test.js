import {renderWithRedux} from "../../../tests/helpers/renderWithRedux";
import DiskButtons from "../DiskButtons/DiskButtons";
import {screen} from "@testing-library/react";
import File from "./File";
import userEvent from "@testing-library/user-event";

describe('DiskButtons', () => {
	const file = {
		_id: "636a83931703b3c87bb4e834",
		name: "123.txt",
		type: "txt",
		size: 546,
		path: "123",
		user: "6345c4d37f5b72d7cde90dc4",
		childs: [],
		date: "Tue Nov 08 2022 18:28:03 GMT+0200 (Восточная Европа, стандартное время)", "__v": 0
	}
	const refetch = () => {
	}

	test('match snapshot', async () => {
		renderWithRedux(<File file={file} refetch={refetch}/>)
		const fileElem = await screen.getByTestId('file-elem')
		expect(fileElem).toBeInTheDocument()
		await userEvent.hover(fileElem)

		const downloadIcon = await screen.getByTestId('download-icon')
		const deleteIcon = await screen.getByTestId('delete-icon')
		const shareIcon = await screen.getByTestId('share-icon')
		expect(downloadIcon).toBeInTheDocument()
		expect(deleteIcon).toBeInTheDocument()
		expect(shareIcon).toBeInTheDocument()
		expect(fileElem).toMatchSnapshot()

	})

});