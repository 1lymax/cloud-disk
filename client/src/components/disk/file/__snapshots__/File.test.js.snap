// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DiskButtons match snapshot 1`] = `
<div
  class="list"
  data-testid="file-elem"
>
  <div
    class="list__icon"
  >
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium list__icon__file css-i4bv87-MuiSvgIcon-root"
      data-testid="file-icon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M14.17 5 19 9.83V19H5V5h9.17m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V9.83c0-.53-.21-1.04-.59-1.41l-4.83-4.83c-.37-.38-.88-.59-1.41-.59zM7 15h10v2H7v-2zm0-4h10v2H7v-2zm0-4h7v2H7V7z"
      />
    </svg>
  </div>
  <div
    class="list__name"
  >
    <h6
      class="MuiTypography-root MuiTypography-subtitle2 css-ati914-MuiTypography-root"
    >
      123.txt
    </h6>
  </div>
  <div
    class="list__actions"
    style=""
  >
    <button
      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorSecondary MuiIconButton-sizeMedium css-1gws2xf-MuiButtonBase-root-MuiIconButton-root"
      data-testid="download-icon"
      tabindex="0"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="CloudDownloadIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M19.35 10.04C18.67 6.59 15.64 4 12 4 9.11 4 6.6 5.64 5.35 8.04 2.34 8.36 0 10.91 0 14c0 3.31 2.69 6 6 6h13c2.76 0 5-2.24 5-5 0-2.64-2.05-4.78-4.65-4.96zM17 13l-5 5-5-5h3V9h4v4h3z"
        />
      </svg>
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
    <button
      class="MuiButtonBase-root MuiButton-root MuiLoadingButton-root MuiButton-text MuiButton-textSecondary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiLoadingButton-root MuiButton-text MuiButton-textSecondary MuiButton-sizeMedium MuiButton-textSizeMedium css-1jpsq0x-MuiButtonBase-root-MuiButton-root-MuiLoadingButton-root"
      data-testid="delete-icon"
      id=":r0:"
      tabindex="0"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="DeleteForeverIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zm2.46-7.12 1.41-1.41L12 12.59l2.12-2.12 1.41 1.41L13.41 14l2.12 2.12-1.41 1.41L12 15.41l-2.12 2.12-1.41-1.41L10.59 14l-2.13-2.12zM15.5 4l-1-1h-5l-1 1H5v2h14V4z"
        />
      </svg>
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
    <button
      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorSecondary MuiIconButton-sizeMedium css-1gws2xf-MuiButtonBase-root-MuiIconButton-root"
      data-testid="share-icon"
      tabindex="0"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="ShareIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M18 16.08c-.76 0-1.44.3-1.96.77L8.91 12.7c.05-.23.09-.46.09-.7s-.04-.47-.09-.7l7.05-4.11c.54.5 1.25.81 2.04.81 1.66 0 3-1.34 3-3s-1.34-3-3-3-3 1.34-3 3c0 .24.04.47.09.7L8.04 9.81C7.5 9.31 6.79 9 6 9c-1.66 0-3 1.34-3 3s1.34 3 3 3c.79 0 1.5-.31 2.04-.81l7.12 4.16c-.05.21-.08.43-.08.65 0 1.61 1.31 2.92 2.92 2.92 1.61 0 2.92-1.31 2.92-2.92s-1.31-2.92-2.92-2.92z"
        />
      </svg>
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </div>
  <div
    class="list__date"
  >
    <h6
      aria-label="Nov 08 2022  18:28:03"
      class="MuiTypography-root MuiTypography-subtitle2 css-ati914-MuiTypography-root"
      data-mui-internal-clone-element="true"
    >
      Nov 08 2022
    </h6>
  </div>
  <div
    class="list__size"
  >
    <h6
      class="MuiTypography-root MuiTypography-subtitle2 css-ati914-MuiTypography-root"
    >
      546B
    </h6>
  </div>
</div>
`;
